File: ios/LookSeeProto/UI Files/LandmarkScan.swift
//
//  LandmarkScan.swift
//  LookSeeProto
//

import SwiftUI
import CoreLocation

struct LandmarkScan: View {
    var onTap: () -> Void = {}
    var onPinch: () -> Void = {}

    @Binding var isDetecting: Bool
    @Binding var isNavVisible: Bool
    
    var isActive: Bool = true
    
    @StateObject private var detector = Detector()
    @ObservedObject private var infoView = VariableContainer.shared
    @EnvironmentObject var vm: AuthViewModel
    
    @State private var zoomLevel: CGFloat = 1.0
    @State private var zoomIndicatorVisible = false
    @State private var zoomFadeTask: Task<Void, Never>?
    @State private var liveInfoFetchTask: Task<Void, Never>?

    @State private var isCameraPaused = false
    @State private var showThresholdControls = false
    @State private var isWarmingUp = true // 🚀 NEW: Masks the camera snap

    var body: some View {
        GeometryReader { geo in
            let lockedSafeZone = CGRect(
                x: geo.size.width * 0.15,
                y: geo.size.height * 0.20,
                width: geo.size.width * 0.70,
                height: geo.size.height * 0.45
            )

            ZStack(alignment: .center) {
                let blurAmount = infoView.infoView ? 10.0 : 0.0

                CameraPreview(
                    detector: detector,
                    zoomLevel: $zoomLevel,
                    showSafeZone: .constant(false),
                    safeZoneRect: .constant(lockedSafeZone),
                    onTap: onTap,
                    onPinch: onPinch,
                    isAIPaused: $isCameraPaused,
                    onBoxTap: { detection in
                        openPopup(for: detection)
                    }
                )
                .ignoresSafeArea()
                .blur(radius: blurAmount)
                .onChange(of: zoomLevel) { _, _ in
                    showZoomIndicatorThenFade()
                    onTap()
                }
                .onChange(of: detector.currentLabel) { _, newLabel in
                    withAnimation(.easeOut(duration: 0.1)) {
                        isDetecting = isActive &&
                            newLabel?.trimmingCharacters(
                                in: .whitespacesAndNewlines
                            ).isEmpty == false
                    }
                }

                if !isActive {
                    Color.black
                        .ignoresSafeArea()
                        .zIndex(2)
                }
                
                // 🚀 NEW: Solid black overlay that hides the physical lens snapping
                if isWarmingUp {
                    Color.black
                        .ignoresSafeArea()
                        .zIndex(8)
                        .transition(.opacity)
                }
                
                // --- Confidence Slider ---
                if isActive && !infoView.infoView {
                    VStack {
                        Spacer()
                        HStack {
                            Spacer()
                            Button {
                                withAnimation(.easeInOut(duration: 0.25)) {
                                    showThresholdControls.toggle()
                                }
                            } label: {
                                Image(systemName: showThresholdControls ? "xmark.circle.fill" : "slider.horizontal.3")
                                    .font(.system(size: 22))
                                    .foregroundStyle(.white)
                                    .padding(10)
                                    .background(Color.black.opacity(0.6), in: Circle())
                            }
                            .padding(.trailing, 16)
                            .padding(.bottom, 16)
                        }
                    }
                    .zIndex(7)
                }
                
                if isActive && !infoView.infoView && showThresholdControls {
                    HStack {
                        Spacer()
                        VStack(spacing: 16) {
                            HStack {
                                Spacer()
                                Button {
                                    withAnimation(.easeInOut(duration: 0.25)) {
                                        showThresholdControls = false
                                    }
                                } label: {
                                    Image(systemName: "xmark.circle.fill")
                                        .foregroundStyle(.white.opacity(0.7))
                                }
                            }

                            Text("\(Int(detector.confidenceThreshold * 100))% - \(Int(detector.confidenceThreshold * detector.ThresholdRangemultiplier * 100))%")
                                .font(.caption2.monospacedDigit())
                                .fontWeight(.bold)
                                .foregroundStyle(.white)

                            Slider(value: $detector.confidenceThreshold, in: 0.1...0.95, step: 0.05)
                                .tint(.green)
                                .frame(width: 120, height: 20)
                                .rotationEffect(.degrees(-90))
                                .frame(width: 20, height: 120)

                            Text("detector threshold (0.1-0.95) \(Int(detector.confidenceThreshold * 100))%")
                                .font(.system(size: 10))

                            Slider(value: $detector.ThresholdRangemultiplier, in: 0.1...1.0, step: 0.05)
                                .tint(.green)
                                .frame(width: 120, height: 20)
                                .rotationEffect(.degrees(-90))
                                .frame(width: 20, height: 120)

                            Text("detector range multiplier (0.1-1.0) \((detector.ThresholdRangemultiplier))")
                                .font(.system(size: 10))
                        }
                        .padding(.vertical, 16)
                        .padding(.horizontal, 12)
                        .background(Color.black.opacity(0.6), in: RoundedRectangle(cornerRadius: 12))
                        .padding(.trailing, 16)
                        .padding(.bottom, 120)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomTrailing)
                    .zIndex(6)
                    .transition(.move(edge: .trailing).combined(with: .opacity))
                }

                if isActive,
                   !infoView.infoView,
                   zoomIndicatorVisible {
                    VStack {
                        Spacer()

                        Text(String(format: "%.1fx", zoomLevel))
                            .font(.caption.monospacedDigit())
                            .fontWeight(.bold)
                            .foregroundStyle(.white)
                            .padding(.horizontal, 16)
                            .padding(.vertical, 8)
                            .background(
                                Color.black.opacity(0.6),
                                in: RoundedRectangle(cornerRadius: 12)
                            )
                            .padding(.bottom, 110)
                            .transition(.opacity)
                    }
                    .zIndex(5)
                }
            }
            .animation(
                .easeOut(duration: 0.25),
                value: zoomIndicatorVisible
            )
            .onAppear {
                detector.dynamicSafeZone = lockedSafeZone
                detector.hideBoundingBoxes = false
                updatePauseState()

                // 🚀 Mask the camera snapping effect during initialization
                isWarmingUp = true
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.65) {
                    withAnimation(.easeOut(duration: 0.3)) {
                        isWarmingUp = false
                    }
                }
            }
            .onChange(of: geo.size) { _, _ in
                detector.dynamicSafeZone = lockedSafeZone
            }
            .onChange(of: isActive) { _, active in
                updatePauseState()
                if active {
                    isWarmingUp = true
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.65) {
                        withAnimation(.easeOut(duration: 0.3)) {
                            isWarmingUp = false
                        }
                    }
                }
            }
            .onChange(of: infoView.infoView) { _, _ in
                updatePauseState()
            }
            .onDisappear {
                liveInfoFetchTask?.cancel()
                zoomFadeTask?.cancel()
                isCameraPaused = true
                isDetecting = false
            }
        }
    }

    private func getCityName(latitude: Double, longitude: Double) async -> String? {
        let location = CLLocation(latitude: latitude, longitude: longitude)
        let geocoder = CLGeocoder()
        do {
            let placemarks = try await geocoder.reverseGeocodeLocation(location)
            if let place = placemarks.first {
                let city = place.locality ?? ""
                let state = place.administrativeArea ?? ""
                if !city.isEmpty && !state.isEmpty { return "\(city), \(state)" }
                return city.isEmpty ? state : city
            }
        } catch {
            print("Geocoding error: \(error)")
        }
        return nil
    }

    // MARK: - Internal Methods
    private func openPopup(for detection: Detection) {
        liveInfoFetchTask?.cancel()

        guard let entry = detection.landmarkEntry else {
            infoView.landmarkId = ""
            infoView.landmarkName = detection.displayLabel
            infoView.landmarkConfidence = detection.confidence * 100
            infoView.landmarkDescription = "Discover more about this location."
            infoView.landmarkURL = ""
            infoView.landmarkWebsiteUrl = ""
            infoView.promoName = "No active promotion"
            infoView.promoDescription = ""
            infoView.promoImageUrl = ""
            infoView.infoView = true

            return
        }

        infoView.presentLandmark(
            entry,
            clusterId: Int(detection.clusterID) ?? 0,
            trainingRunId: detection.modelVersion,
            detectionConfidence: detection.confidence
        )

        if vm.tier == "business" || vm.hasActiveSubscription {
            let lat = entry.latitude ?? 0.0
            let lon = entry.longitude ?? 0.0
            let displayLabel = detection.displayLabel
            let lId = entry.landmarkId
            let cachedImg = infoView.merchantLogoUrl

            Task {
                let cityString = await getCityName(latitude: lat, longitude: lon) ?? "Unknown Location"
                await vm.logScanHistory(landmarkId: lId, label: displayLabel, location: cityString, latitude: lat, longitude: lon, imageUrl: cachedImg)
            }
        }

        let landmarkId = entry.landmarkId.trimmingCharacters(in: .whitespacesAndNewlines)

        if landmarkId.isEmpty {
            infoView.landmarkWebsiteUrl = ""
            infoView.promoName = "No active promotion"
            infoView.promoDescription = ""
            infoView.promoImageUrl = ""
        } else {
            fetchLiveLandmarkInfo(for: landmarkId)
        }
    }

    private func fetchLiveLandmarkInfo(for landmarkId: String) {
        liveInfoFetchTask?.cancel()

        liveInfoFetchTask = Task {
            do {
                let liveInfo = try await LiveLandmarkInfoService()
                    .fetchLiveInfo(
                        landmarkId: landmarkId,
                        timeoutSeconds: 2.5
                    )

                guard !Task.isCancelled else { return }

                await MainActor.run {
                    guard infoView.landmarkId == landmarkId else { return }
                    applyLiveInfo(liveInfo, landmarkId: landmarkId)
                }
            } catch {
                guard !Task.isCancelled else { return }

                await MainActor.run {
                    guard infoView.landmarkId == landmarkId else { return }
                    print("⚠️ Live landmark info unavailable. Keeping manifest fallback.")
                }
            }
        }
    }

    @MainActor
    private func applyLiveInfo(_ liveInfo: LiveLandmarkInfoResponse, landmarkId: String) {
        let liveLabel = liveInfo.label.trimmingCharacters(in: .whitespacesAndNewlines)
        let liveDescription = liveInfo.shortDescription.trimmingCharacters(in: .whitespacesAndNewlines)
        let liveWebsiteUrl = liveInfo.websiteUrl?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        if !liveLabel.isEmpty { infoView.landmarkName = liveLabel }
        if !liveDescription.isEmpty { infoView.landmarkDescription = liveDescription }
        infoView.landmarkWebsiteUrl = liveWebsiteUrl

        if liveInfo.isActive == false {
            infoView.promoName = "No active promotion"
            infoView.promoDescription = ""
            infoView.promoImageUrl = ""
            return
        }

        if let promotion = liveInfo.activePromotion {
            let promoName = promotion.name.trimmingCharacters(in: .whitespacesAndNewlines)
            let promoDescription = promotion.description.trimmingCharacters(in: .whitespacesAndNewlines)
            let promoImageUrl = promotion.imageUrl?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

            if !promoName.isEmpty {
                infoView.promoName = promoName
                infoView.promoDescription = promoDescription
                infoView.promoImageUrl = promoImageUrl
            } else {
                infoView.promoName = "No active promotion"
                infoView.promoDescription = ""
                infoView.promoImageUrl = ""
            }
        } else {
            infoView.promoName = "No active promotion"
            infoView.promoDescription = ""
            infoView.promoImageUrl = ""
        }
    }

    private func updatePauseState() {
        isCameraPaused = !isActive
        if !isActive { isDetecting = false }
    }

    private func showZoomIndicatorThenFade() {
        zoomFadeTask?.cancel()
        zoomIndicatorVisible = true

        zoomFadeTask = Task {
            try? await Task.sleep(nanoseconds: 1_200_000_000)
            guard !Task.isCancelled else { return }

            await MainActor.run {
                withAnimation(.easeOut(duration: 0.25)) {
                    zoomIndicatorVisible = false
                }
            }
        }
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/uifiles/LandmarkScan.kt
package looksee.angelll.com.uifiles

import android.content.Context
import android.location.Geocoder
import android.util.Log
import androidx.compose.animation.*
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import looksee.angelll.com.detection.*
import looksee.angelll.com.models.*
import looksee.angelll.com.services.*
import java.util.Locale
import kotlin.time.Duration.Companion.milliseconds

private suspend fun getCityName(context: Context, lat: Double, lon: Double): String? = withContext(Dispatchers.IO) {
    try {
        val geocoder = Geocoder(context, Locale.getDefault())
        val addresses = geocoder.getFromLocation(lat, lon, 1)
        if (!addresses.isNullOrEmpty()) {
            val city = addresses[0].locality ?: ""
            val state = addresses[0].adminArea ?: ""
            if (city.isNotEmpty() && state.isNotEmpty()) return@withContext "$city, $state"
            if (city.isNotEmpty() || state.isNotEmpty()) return@withContext city.ifEmpty { state }
        }
    } catch (e: Exception) {
        Log.e("Geocoder", "Error: ${e.message}")
    }
    return@withContext String.format(Locale.US, "%.5f, %.5f", lat, lon)
}

@Composable
fun LandmarkScan(
    vm: looksee.angelll.com.viewmodels.AuthViewModel? = null,
    onTap: () -> Unit = {},
    onPinch: () -> Unit = {},
    isDetecting: Boolean,
    onIsDetectingChange: (Boolean) -> Unit,
    isNavVisible: Boolean,
    isActive: Boolean = true
) {
    val context = LocalContext.current
    val haptic = LocalHapticFeedback.current
    val coroutineScope = rememberCoroutineScope()
    val infoView = remember { VariableContainer.shared }
    val detector = remember { Detector.shared(context) }

    var isPopupOpen by remember { mutableStateOf(infoView.infoView) }
    LaunchedEffect(Unit) {
        while (true) {
            if (isPopupOpen != infoView.infoView) {
                isPopupOpen = infoView.infoView
            }
            delay(100)
        }
    }

    var zoomLevel by remember { mutableStateOf(1f) }
    var zoomIndicatorVisible by remember { mutableStateOf(false) }
    var zoomFadeJob by remember { mutableStateOf<Job?>(null) }

    var isCameraPaused by remember { mutableStateOf(false) }
    var showThresholdControls by remember { mutableStateOf(false) }

    var liveInfoFetchJob by remember { mutableStateOf<Job?>(null) }

    var confidenceThreshold by remember { mutableStateOf(detector.confidenceThreshold) }
    var thresholdMultiplier by remember { mutableStateOf(detector.trackingThresholdMultiplier) }

    fun updatePauseState() {
        isCameraPaused = !isActive
        if (!isActive) onIsDetectingChange(false)
    }

    fun showZoomIndicatorThenFade() {
        zoomFadeJob?.cancel()
        zoomIndicatorVisible = true
        zoomFadeJob = coroutineScope.launch {
            delay(1200.milliseconds)
            zoomIndicatorVisible = false
        }
    }

    fun fetchLiveLandmarkInfo(landmarkId: String, lat: Double, lon: Double, label: String) {
        liveInfoFetchJob?.cancel()
        liveInfoFetchJob = coroutineScope.launch {
            try {
                val liveInfo = LiveLandmarkInfoService(context).fetchLiveInfo(landmarkId, 2.5)
                if (infoView.landmarkId == landmarkId) {
                    applyLiveInfo(liveInfo, infoView)

                    val realAddress = liveInfo.merchantAddress?.takeIf { it.isNotBlank() }
                    if (realAddress != null && vm != null && vm.hasActiveSubscription) {
                        vm.logScanHistory(
                            landmarkId = landmarkId,
                            label = liveInfo.label.ifBlank { label },
                            location = realAddress,
                            latitude = lat,
                            longitude = lon,
                            imageUrl = liveInfo.merchantLogoUrl ?: infoView.merchantLogoUrl
                        )
                    }
                }
            } catch (e: Exception) {
                Log.e("LandmarkScan", "Live info unavailable: ${e.message}")
            }
        }
    }

    fun openPopup(detection: Detection) {
        liveInfoFetchJob?.cancel()
        val entry = detection.landmarkEntry()

        infoView.infoView = true
        isPopupOpen = true

        if (entry == null) {
            infoView.resetLandmarkDisplay()
            infoView.landmarkName = detection.displayLabel()
            infoView.landmarkConfidence = detection.confidence * 100
            infoView.landmarkDescription = "Discover more about this location."
            return
        }

        infoView.presentLandmark(
            entry = entry,
            clusterId = detection.clusterId.toIntOrNull() ?: 0,
            trainingRunId = detection.modelVersion,
            detectionConfidence = detection.confidence
        )

        val lat = entry.latitude
        val lon = entry.longitude
        val displayLabel = detection.displayLabel()
        val lId = entry.landmarkId
        val cachedImg = infoView.merchantLogoUrl

        if (vm != null && vm.hasActiveSubscription) {
            coroutineScope.launch {
                val locationFallback = getCityName(context, lat, lon) ?: "Current Location"

                vm.logScanHistory(
                    landmarkId = lId,
                    label = displayLabel,
                    location = locationFallback,
                    latitude = lat,
                    longitude = lon,
                    imageUrl = cachedImg
                )
            }
        }

        if (entry.landmarkId.isNotBlank()) {
            fetchLiveLandmarkInfo(entry.landmarkId, lat, lon, displayLabel)
        }
    }

    LaunchedEffect(isActive) {
        updatePauseState()
    }

    val currentLabel by detector.currentLabel.collectAsState()
    LaunchedEffect(currentLabel, isActive) {
        onIsDetectingChange(isActive && !currentLabel.isNullOrBlank())
    }

    val newlyDetectedLandmark by detector.newlyDetectedLandmark.collectAsState()
    LaunchedEffect(newlyDetectedLandmark) {
        if (newlyDetectedLandmark != null) {
            haptic.performHapticFeedback(HapticFeedbackType.LongPress)
        }
    }

    BoxWithConstraints(modifier = Modifier.fillMaxSize()) {
        val width = maxWidth
        val height = maxHeight

        val lockedSafeZone = remember(width, height) {
            DetectionBox(
                left = (width.value * 0.15f),
                top = (height.value * 0.20f),
                right = (width.value * 0.85f),
                bottom = (height.value * 0.65f)
            )
        }

        ZStack(alignment = Alignment.Center) {

            CameraPreview(
                detector = detector,
                zoomLevel = zoomLevel,
                onZoomLevelChange = {
                    zoomLevel = it
                    showZoomIndicatorThenFade()
                    onTap()
                },
                showSafeZone = false,
                safeZoneRect = lockedSafeZone,
                onTap = onTap,
                onPinch = onPinch,
                isAIPaused = isCameraPaused || isPopupOpen,
                onBoxTap = { openPopup(it) },
                hideBoundingBoxes = isPopupOpen,
                modifier = Modifier.fillMaxSize()
            )

            if (infoView.infoView) {
                Box(modifier = Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.5f)).zIndex(1f))
            }

            if (!isActive) {
                Box(modifier = Modifier.fillMaxSize().background(Color.Black).zIndex(2f))
            }

            if (isActive && !isPopupOpen) {
                Box(modifier = Modifier.fillMaxSize().padding(bottom = 120.dp, end = 16.dp), contentAlignment = Alignment.BottomEnd) {
                    IconButton(
                        onClick = { showThresholdControls = !showThresholdControls },
                        modifier = Modifier.background(Color.Black.copy(0.6f), CircleShape)
                    ) {
                        Icon(
                            imageVector = if (showThresholdControls) Icons.Default.Close else Icons.Default.Tune,
                            contentDescription = null,
                            tint = Color.White
                        )
                    }
                }
            }

            AnimatedVisibility(
                visible = isActive && !isPopupOpen && showThresholdControls,
                enter = slideInHorizontally { it } + fadeIn(),
                exit = slideOutHorizontally { it } + fadeOut()
            ) {
                Box(modifier = Modifier.fillMaxSize().padding(bottom = 180.dp, end = 16.dp), contentAlignment = Alignment.BottomEnd) {
                    Surface(
                        color = Color.Black.copy(0.7f),
                        shape = RoundedCornerShape(16.dp),
                        modifier = Modifier.width(180.dp)
                    ) {
                        Column(modifier = Modifier.padding(16.dp), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(12.dp)) {
                            Text(
                                text = "${(confidenceThreshold * 100).toInt()}% - ${(confidenceThreshold * thresholdMultiplier * 100).toInt()}%",
                                color = Color.White, fontSize = 12.sp, fontWeight = FontWeight.Bold
                            )

                            Slider(
                                value = confidenceThreshold,
                                onValueChange = {
                                    confidenceThreshold = it
                                    detector.confidenceThreshold = it
                                },
                                valueRange = 0.1f..0.95f,
                                colors = SliderDefaults.colors(thumbColor = Color.Green, activeTrackColor = Color.Green)
                            )
                            Text("Threshold", color = Color.White, fontSize = 10.sp)

                            Slider(
                                value = thresholdMultiplier,
                                onValueChange = {
                                    thresholdMultiplier = it
                                    detector.trackingThresholdMultiplier = it
                                },
                                valueRange = 0.1f..1.0f,
                                colors = SliderDefaults.colors(thumbColor = Color.Green, activeTrackColor = Color.Green)
                            )
                            Text("Range Multiplier", color = Color.White, fontSize = 10.sp)
                        }
                    }
                }
            }

            if (isActive && !isPopupOpen && zoomIndicatorVisible) {
                Box(modifier = Modifier.fillMaxSize().padding(bottom = 110.dp), contentAlignment = Alignment.BottomCenter) {
                    Surface(
                        color = Color.Black.copy(0.6f),
                        shape = RoundedCornerShape(12.dp)
                    ) {
                        Text(
                            text = String.format(Locale.US, "%.1fx", zoomLevel),
                            color = Color.White,
                            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
            }
        }
    }

    DisposableEffect(Unit) {
        detector.setHideBoundingBoxes(false)
        onDispose {
            liveInfoFetchJob?.cancel()
            zoomFadeJob?.cancel()
            isCameraPaused = true
            onIsDetectingChange(false)
        }
    }
}

private fun applyLiveInfo(liveInfo: LiveLandmarkInfoResponse, infoView: VariableContainer) {
    if (liveInfo.label.isNotBlank()) infoView.landmarkName = liveInfo.label
    if (liveInfo.shortDescription.isNotBlank()) infoView.landmarkDescription = liveInfo.shortDescription
    infoView.landmarkWebsiteUrl = liveInfo.websiteUrl ?: ""

    if (liveInfo.isActive == false) {
        infoView.promoName = "No active promotion"
        return
    }

    liveInfo.activePromotion?.let { promo ->
        infoView.promoName = promo.name
        infoView.promoDescription = promo.description
        infoView.promoImageUrl = promo.imageUrl ?: ""
    } ?: run {
        infoView.promoName = "No active promotion"
    }
}

@Composable
fun ZStack(
    modifier: Modifier = Modifier,
    alignment: Alignment = Alignment.Center,
    content: @Composable (BoxScope.() -> Unit)
) {
    Box(modifier = modifier, contentAlignment = alignment, content = content)
}